src/metric.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+//! The purpose of this module is to provide abstractions for working with
+//! metrics in the context of rust-vmm components where there is a strong need
+//! to have metrics as an optional feature.
+//!
+//! As multiple stakeholders are using these components, there are also
+//! questions regarding the serialization format, as metrics are expected to be
+//! flexible enough to allow different formatting, serialization and writers.
+//! When using the rust-vmm metrics, the expectation is that VMMs built on top
+//! of these components can choose what metrics they’re interested in and also
+//! can add their own custom metrics without the need to maintain forks.
+
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+/// Abstraction over the common metric operations.
+///
+/// An object implementing `Metric` is expected to have an inner counter that
+/// can be incremented and reset. The `Metric` trait can be used for
+/// implementing a metric system backend (or an aggregator).
+pub trait Metric {
```

Comment:
How about adding an
```
fn set(&self, value);
```
for non-accumulate counters?

Reply:
Sounds good. I added it.